### PR TITLE
Revise driver/extension terminology in ext-mongodb docs

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPLIB-1279

### CODE REVIEW

File: language-snippets.ent [hunk 1]
```diff
@@ -3573,7 +3573,7 @@ local: {
 <!ENTITY mongodb.note.forking '
    <note xmlns="http://docbook.org/ns/docbook">
     <simpara>
-     On Unix platforms, the MongoDB driver is sensitive to scripts that use the
+     On Unix platforms, the extension is sensitive to scripts that use the
```
Comment (jmikola) [inline language-snippets.ent:3576]: Note: I did leave various references to "driver" in the docs when the text was referring to the concept of a driver. A common example is in the copied URI option or CSFLE docs, which talk about driver behavior.

This is an instance where we're very much discussing the extension.

File: reference/mongodb/bson.xml [hunk 2]
```diff
@@ -2,7 +2,7 @@
 <!-- $Revision$ -->
 
  <book xml:id="book.bson" xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink">
-  <title>BSON type classes and serialization functions</title>
+  <title>MongoDB Extension BSON Classes and Functions</title>
```
Comment (GromNaN) [inline reference/mongodb/bson.xml:5]: I don't thing you need to explicitly tell this classes come from the extension.
```suggestion
  <title>MongoDB BSON Classes and Functions</title>
```

That matches the namespace `MongoDB\BSON`.

File: reference/mongodb/bson/serializable/bsonserialize.xml [hunk 3]
```diff
@@ -30,8 +30,8 @@
   <para>
    Users are encouraged to include an <property>_id</property> property (e.g. a
    <classname>MongoDB\BSON\ObjectId</classname> initialized in your constructor)
-   when returning data for a BSON root document; otherwise, the driver or
-   database will need to generate a
+   when returning data for a BSON root document; otherwise, the extension or
+   server will need to generate a
```
Comment (GromNaN) [inline reference/mongodb/bson/serializable/bsonserialize.xml:34]:
```suggestion
   the server will need to generate a
```

Which component actually generates the ObjectId?

File: reference/mongodb/mongodb/driver/monitoring/logsubscriber/log.xml [hunk 4]
```diff
@@ -16,8 +16,8 @@
    <methodparam><type>string</type><parameter>message</parameter></methodparam>
   </methodsynopsis>
   <para>
-   If the subscriber has been registered, the driver will call this method for
-   each logged message.
+   If the subscriber has been registered, this method will be called for each
+   logged message.
```
Comment (GromNaN) [inline reference/mongodb/mongodb/driver/monitoring/logsubscriber/log.xml:20]: Why using the future? Present is affirmative:

```suggestion
   If the subscriber is registered, this method is called for each logged message.
```

File: reference/mongodb/tutorial/library.xml [hunk 5]
```diff
@@ -147,11 +146,11 @@ foreach ($result as $entry) {
     While it may not be apparent in the examples, BSON documents and arrays are
     unserialized as type classes in the library by default. These classes ensure
     that values preserve their type when being serialized back into BSON, which
-    avoids a caveat in the old driver where arrays might turn into documents,
-    and vice versa. Additionally, these type classes extend
-    <classname>ArrayObject</classname> for enhanced usability. You can find more
-    information on how serialization and deserialization between PHP variables
-    and BSON is handled by the driver and library by reading the
+    avoids a caveat in the legacy <code>mongo</code> extension where arrays
```
Comment (GromNaN) [inline reference/mongodb/tutorial/library.xml:149]: I wondered if it was worth continuing to mention the legacy `mongo` extension, but it serves the purpose of explaining why these classes exist.

The concept of "type classes" is not clear for me when I first read this paragraph. Know I understand that's BSON classes.

File: reference/mongodb/bson.xml (hunk 2 above)
Comment (Girgias) [reply reference/mongodb/bson.xml:5]: I agree with the suggestion above.

File: reference/mongodb/bson/binary.xml [hunk 6]
```diff
@@ -154,7 +154,7 @@
      <term><constant>MongoDB\BSON\Binary::TYPE_OLD_UUID</constant></term>
      <listitem>
       <para>Universally unique identifier (deprecated in favor of <constant>MongoDB\BSON\Binary::TYPE_UUID</constant>). When using this type, the Binary&apos;s data should be 16 bytes in length.</para>
-      <para>Historically, other drivers encoded values with this type based on their language conventions (e.g. varying endianness), which makes it non-portable. The PHP driver applies no special handling for encoding or decoding data with this type.</para>
+      <para>Historically, other drivers encoded values with this type based on their language conventions (e.g. varying endianness), which makes it non-portable. The PHP extension applies no special handling for encoding or decoding data with this type.</para>
```
Comment (Girgias) [inline reference/mongodb/bson/binary.xml:157]: Could you add line breaks while you are at it?

File: reference/mongodb/bson/serializable/bsonserialize.xml [hunk 7]
```diff
@@ -30,8 +30,8 @@
   <para>
    Users are encouraged to include an <property>_id</property> property (e.g. a
    <classname>MongoDB\BSON\ObjectId</classname> initialized in your constructor)
```
Comment (Girgias) [inline reference/mongodb/bson/serializable/bsonserialize.xml:32]: Drive-by comment that has nothing to do with the PR, but this sentence uses personalization (i.e. using `you`) which we avoid in the manual.

File: reference/mongodb/mongodb/driver/manager/construct.xml [hunk 8]
```diff
@@ -1271,8 +1271,8 @@ mongodb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][
          <literal>"connectTimeoutMS"</literal> URI option now defaults to 10
          seconds instead of
          <link linkend="ini.default-socket-timeout">default_socket_timeout</link>
-         in previous versions. Additionally, the driver no longer supports all
-         <link linkend="context.ssl">SSL context options</link> via the
+         in previous versions. Additionally, the extension no longer supports
+         all <link linkend="context.ssl">SSL context options</link> via the
```
Comment (Girgias) [inline reference/mongodb/mongodb/driver/manager/construct.xml:1275]: Is it me or the usage of `all` here sounds unatural?

**Review by Girgias — COMMENTED** [review]
Had a quick look and looks fine to me, just had some remarks.

**Review by alcaeus — COMMENTED** [review]
Changes LGTM pending changes suggested by @GromNaN and @Girgias.

File: reference/mongodb/bson/serializable/bsonserialize.xml (hunk 3 above)
Comment (jmikola) [reply reference/mongodb/bson/serializable/bsonserialize.xml:34]: Both the driver and server have logic to add a missing ObjectId, but you'd have to manually execute an `insert` command in order to bypass the driver logic and put the onus on the server.

I think the original grammar was correct here ("the" can apply to both), but it's probably easier to just remove mention of the server here.

Comment (jmikola) [reply reference/mongodb/bson/serializable/bsonserialize.xml:34]: Oh, I responded without considering the full context of this line. The sentence here _is_ correctly stating that the extension will generate the `_id` on insert, and the server will do so on upsert. Nevertheless, I can reword this to be more clear.

Comment (jmikola) [reply reference/mongodb/bson/serializable/bsonserialize.xml:34]: I'm rewriting this paragraph to be more clear.

File: reference/mongodb/mongodb/driver/monitoring/logsubscriber/log.xml (hunk 4 above)
Comment (jmikola) [reply reference/mongodb/mongodb/driver/monitoring/logsubscriber/log.xml:20]: Rewrote all of these sentences to use present tense.

File: reference/mongodb/tutorial/library.xml (hunk 5 above)
Comment (jmikola) [reply reference/mongodb/tutorial/library.xml:149]: I rewrote this to remove "type classes" and explicitly mention what interfaces are being implemented.

File: reference/mongodb/mongodb/driver/manager/construct.xml (hunk 8 above)
Comment (jmikola) [reply reference/mongodb/mongodb/driver/manager/construct.xml:1275]: I think "all" is correct here. When the extension originally used PHP streams, literally all options would be honored since PHP enforced them. When we stopped using PHP streams, we added logic to parse _some_ stream options from a stream context (only those that mapped to something in libmongoc).

File: reference/mongodb/bson/serializable/bsonserialize.xml (hunk 7 above)
Comment (jmikola) [reply reference/mongodb/bson/serializable/bsonserialize.xml:32]: I removed "your" while rewriting this paragraph, but there are still plenty more style violations in our docs. I've opened [PHPC-2393](https://jira.mongodb.org/browse/PHPC-2393) to track that.

File: reference/mongodb/bson/binary.xml (hunk 6 above)
Comment (jmikola) [reply reference/mongodb/bson/binary.xml]: Added a commit to wrap long lines in this file.

File: reference/mongodb/bson/serializable/bsonserialize.xml (hunk 7 above)
Comment (Girgias) [reply reference/mongodb/bson/serializable/bsonserialize.xml]: ACK

File: reference/mongodb/bson/binary.xml (hunk 6 above)
Comment (Girgias) [reply reference/mongodb/bson/binary.xml]: Thank you!

File: reference/mongodb/mongodb/driver/manager/construct.xml (hunk 8 above)
Comment (Girgias) [reply reference/mongodb/mongodb/driver/manager/construct.xml:1275]: Right, that does look correct indeed.

File: reference/mongodb/tutorial/library.xml [hunk 9]
```diff
@@ -145,14 +144,15 @@ foreach ($result as $entry) {
...
-    unserialized as type classes in the library by default. These classes ensure
-    that values preserve their type when being serialized back into BSON, which
-    avoids a caveat in the old driver where arrays might turn into documents,
-    and vice versa. Additionally, these type classes extend
-    <classname>ArrayObject</classname> for enhanced usability. You can find more
-    information on how serialization and deserialization between PHP variables
-    and BSON is handled by the driver and library by reading the
-    <xref linkend="mongodb.persistence"/> specification.
+    unserialized as special classes in the library by default. These classes
+    extend <classname>ArrayObject</classname> for usability and implement the
```
Comment (Girgias) [inline reference/mongodb/tutorial/library.xml:148]: Unrelated, but ArrayObject has some issues in how it is implemented which I've described in a draft RFC: https://github.com/Girgias/php-rfcs/blob/master/container-offset-behaviour.md

I personally would recommend moving away from extending it, but I can understand the BC issues that may pose.

**Review by Girgias — APPROVED** [review]
LGTM from my side

File: reference/mongodb/tutorial/library.xml (hunk 9 above)
Comment (jmikola) [reply reference/mongodb/tutorial/library.xml:148]: I'm not sure what our options are for abandoning ArrayObject, but I'll keep this in mind. Thanks for sharing.

**Review by bisht2050 — COMMENTED** [review]
LGTM
